src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -240,9 +249,8 @@ public void Set(int index, bool value)
         =========================================================================*/
         public void SetAll(bool value)
         {
-            int fillValue = value ? unchecked(((int)0xffffffff)) : 0;
-            int ints = GetArrayLength(m_length, BitsPerInt32);
-            for (int i = 0; i < ints; i++)
+            int fillValue = value ? -1 : 0;
+            for (int i = 0; i < m_array.Length; i++)
```

Comment:
`Span.Fill`?

Reply:
It isn't great for small inputs (< 32 bytes, i.e. < 4 int array).

Reply:
The ctor was changed to use Fill(-1).  Should that be reverted, or are the cases different somehow?

Reply:
> or are the cases different somehow?

They are different (detailed below) but I had also made a mistake in benchmarking.

Previously, I was using a different scale (i.e. int array/bit length) for the ctor versus SetAll.
For SetAll, we are clearing an int array, so I was testing clearing an integer array constructed from a **byte** array of length 1-16 (so the int array is 1/4th the size).
For the ctor, we are accepting a bit length. Here, I was using **bit** length of 1 - 16 (i.e. int array length of 0). Also, since the ctor does more work, and I had improved its performance, the cost of using Fill (for small lengths, like int array of length 0) was hidden by the improvements. I wasn't measuring just using Fill versus a loop in isolation.

Fixing those and using matching **bit** length for both, gives us a more accurate picture.

1) The break-even point for using Fill versus looping through an int array is **length of 5** (_bit length over 128, like 160_).
2) Given the ctor is doing more work (in all cases) rather than just calling Fill, the regression for small arrays isn't that extreme (20% instead of 3x).
3) Since the ctor will usually be called with defaultValue = false, the break even point for that case (when using Fill) is 0 (_bit length of 2_) and not 5 (_bit length of 160_) because we skip calling Fill for it all together (_rely on arrays being 0-initialized_). Therefore, we can and probably **should** use Fill there but **not** in SetAll.

``` ini

BenchmarkDotNet=v0.11.2, OS=Windows 10.0.17763.107 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009706
  [Host]     : .NET Core ? (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.3 (CoreCLR 4.6.26725.06, CoreFX 4.6.26725.05), 64bit RyuJIT


```
|    Method | Size | Value |      Mean |     Error |    StdDev | Ratio | RatioSD |
|---------- |----- |------ |----------:|----------:|----------:|------:|--------:|
| **OldSetAll** |    **1** |  **True** |  **1.404 ns** | **0.0216 ns** | **0.0202 ns** |  **1.00** |    **0.00** |
| NewSetAll |    1 |  True |  4.451 ns | 0.0777 ns | 0.0689 ns |  3.17 |    0.06 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |    **2** |  **True** |  **1.404 ns** | **0.0301 ns** | **0.0282 ns** |  **1.00** |    **0.00** |
| NewSetAll |    2 |  True |  4.466 ns | 0.0518 ns | 0.0459 ns |  3.18 |    0.08 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |    **4** |  **True** |  **1.475 ns** | **0.2423 ns** | **0.2380 ns** |  **1.00** |    **0.00** |
| NewSetAll |    4 |  True |  4.539 ns | 0.1106 ns | 0.1622 ns |  3.17 |    0.45 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |    **8** |  **True** |  **1.368 ns** | **0.0316 ns** | **0.0296 ns** |  **1.00** |    **0.00** |
| NewSetAll |    8 |  True |  4.485 ns | 0.0665 ns | 0.0589 ns |  3.28 |    0.08 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |   **16** |  **True** |  **1.377 ns** | **0.0312 ns** | **0.0292 ns** |  **1.00** |    **0.00** |
| NewSetAll |   16 |  True |  4.492 ns | 0.1706 ns | 0.1596 ns |  3.26 |    0.14 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |   **31** |  **True** |  **1.344 ns** | **0.0190 ns** | **0.0178 ns** |  **1.00** |    **0.00** |
| NewSetAll |   31 |  True |  4.477 ns | 0.0777 ns | 0.0726 ns |  3.33 |    0.06 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |   **32** |  **True** |  **1.426 ns** | **0.0421 ns** | **0.0329 ns** |  **1.00** |    **0.00** |
| NewSetAll |   32 |  True |  4.524 ns | 0.1500 ns | 0.2103 ns |  3.27 |    0.20 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |   **33** |  **True** |  **1.962 ns** | **0.0487 ns** | **0.0455 ns** |  **1.00** |    **0.00** |
| NewSetAll |   33 |  True |  5.332 ns | 0.1101 ns | 0.0976 ns |  2.72 |    0.08 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |   **64** |  **True** |  **1.984 ns** | **0.0376 ns** | **0.0314 ns** |  **1.00** |    **0.00** |
| NewSetAll |   64 |  True |  5.304 ns | 0.0939 ns | 0.0832 ns |  2.68 |    0.07 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |   **96** |  **True** |  **2.585 ns** | **0.0321 ns** | **0.0285 ns** |  **1.00** |    **0.00** |
| NewSetAll |   96 |  True |  5.828 ns | 0.1068 ns | 0.0891 ns |  2.25 |    0.05 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |  **128** |  **True** |  **3.698 ns** | **0.1040 ns** | **0.0972 ns** |  **1.00** |    **0.00** |
| NewSetAll |  128 |  True |  4.986 ns | 0.0504 ns | 0.0471 ns |  1.35 |    0.04 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |  **160** |  **True** |  **4.019 ns** | **0.0648 ns** | **0.0541 ns** |  **1.00** |    **0.00** |
| NewSetAll |  160 |  True |  4.108 ns | 0.0556 ns | 0.0493 ns |  1.02 |    0.02 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |  **192** |  **True** |  **4.650 ns** | **0.1224 ns** | **0.1360 ns** |  **1.00** |    **0.00** |
| NewSetAll |  192 |  True |  4.535 ns | 0.1084 ns | 0.1014 ns |  0.97 |    0.04 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |  **224** |  **True** |  **5.520 ns** | **0.0714 ns** | **0.0668 ns** |  **1.00** |    **0.00** |
| NewSetAll |  224 |  True |  5.286 ns | 0.0657 ns | 0.0583 ns |  0.96 |    0.02 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |  **256** |  **True** |  **6.704 ns** | **0.1432 ns** | **0.1340 ns** |  **1.00** |    **0.00** |
| NewSetAll |  256 |  True |  4.413 ns | 0.0959 ns | 0.0850 ns |  0.66 |    0.02 |
|           |      |       |           |           |           |       |         |
| **OldSetAll** |  **999** |  **True** | **18.784 ns** | **0.2301 ns** | **0.2039 ns** |  **1.00** |    **0.00** |
| NewSetAll |  999 |  True | 11.058 ns | 0.1389 ns | 0.1299 ns |  0.59 |    0.01 |

``` ini

BenchmarkDotNet=v0.11.2, OS=Windows 10.0.17763.107 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-009706
  [Host]     : .NET Core ? (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.3 (CoreCLR 4.6.26725.06, CoreFX 4.6.26725.05), 64bit RyuJIT


```
|        Method | Size | Value |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |
|-------------- |----- |------ |---------:|----------:|----------:|---------:|------:|--------:|
| **OldLengthCtor** |    **1** | **False** | **12.89 ns** | **0.2655 ns** | **0.2483 ns** | **12.76 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    1 | False | 15.62 ns | 0.1487 ns | 0.1318 ns | 15.61 ns |  1.21 |    0.03 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **1** |  **True** | **12.60 ns** | **0.3090 ns** | **0.2580 ns** | **12.56 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    1 |  True | 14.33 ns | 0.0920 ns | 0.0768 ns | 14.33 ns |  1.14 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **2** | **False** | **12.46 ns** | **0.0714 ns** | **0.0558 ns** | **12.46 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    2 | False | 11.99 ns | 0.1825 ns | 0.1708 ns | 11.93 ns |  0.96 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **2** |  **True** | **12.30 ns** | **0.1123 ns** | **0.0938 ns** | **12.32 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    2 |  True | 14.94 ns | 0.3911 ns | 0.7152 ns | 14.93 ns |  1.21 |    0.08 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **4** | **False** | **13.73 ns** | **0.1167 ns** | **0.1034 ns** | **13.73 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    4 | False | 11.94 ns | 0.1372 ns | 0.1071 ns | 11.93 ns |  0.87 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **4** |  **True** | **12.51 ns** | **0.1789 ns** | **0.1586 ns** | **12.49 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    4 |  True | 14.59 ns | 0.2079 ns | 0.1623 ns | 14.56 ns |  1.17 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **8** | **False** | **12.57 ns** | **0.3242 ns** | **0.3329 ns** | **12.52 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    8 | False | 12.23 ns | 0.4297 ns | 0.4220 ns | 12.07 ns |  0.97 |    0.04 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |    **8** |  **True** | **12.57 ns** | **0.0949 ns** | **0.0741 ns** | **12.54 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |    8 |  True | 14.71 ns | 0.3797 ns | 0.3552 ns | 14.57 ns |  1.18 |    0.03 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **16** | **False** | **12.60 ns** | **0.1569 ns** | **0.1391 ns** | **12.61 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   16 | False | 11.84 ns | 0.0988 ns | 0.0825 ns | 11.81 ns |  0.94 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **16** |  **True** | **12.66 ns** | **0.3169 ns** | **0.3650 ns** | **12.55 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   16 |  True | 15.06 ns | 0.0888 ns | 0.0693 ns | 15.04 ns |  1.19 |    0.04 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **31** | **False** | **12.69 ns** | **0.5071 ns** | **0.6594 ns** | **12.51 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   31 | False | 12.00 ns | 0.2381 ns | 0.2111 ns | 11.97 ns |  0.94 |    0.06 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **31** |  **True** | **12.49 ns** | **0.2127 ns** | **0.1990 ns** | **12.42 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   31 |  True | 13.95 ns | 0.1488 ns | 0.1243 ns | 14.00 ns |  1.12 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **32** | **False** | **12.17 ns** | **0.2433 ns** | **0.2275 ns** | **12.08 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   32 | False | 11.54 ns | 0.1014 ns | 0.0899 ns | 11.52 ns |  0.95 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **32** |  **True** | **12.11 ns** | **0.2945 ns** | **0.3025 ns** | **12.01 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   32 |  True | 13.85 ns | 0.1881 ns | 0.1570 ns | 13.85 ns |  1.15 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **33** | **False** | **12.98 ns** | **0.3320 ns** | **0.7218 ns** | **12.66 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   33 | False | 11.50 ns | 0.1636 ns | 0.1450 ns | 11.42 ns |  0.84 |    0.05 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **33** |  **True** | **12.78 ns** | **0.1061 ns** | **0.0940 ns** | **12.80 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   33 |  True | 14.76 ns | 0.2662 ns | 0.2359 ns | 14.73 ns |  1.15 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **64** | **False** | **14.85 ns** | **0.1736 ns** | **0.1450 ns** | **14.84 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   64 | False | 11.57 ns | 0.1515 ns | 0.1343 ns | 11.54 ns |  0.78 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **64** |  **True** | **12.81 ns** | **0.3329 ns** | **0.6949 ns** | **12.53 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   64 |  True | 14.57 ns | 0.1685 ns | 0.1494 ns | 14.56 ns |  1.08 |    0.06 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **96** | **False** | **14.20 ns** | **0.1918 ns** | **0.1794 ns** | **14.20 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   96 | False | 12.04 ns | 0.1951 ns | 0.1825 ns | 11.97 ns |  0.85 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |   **96** |  **True** | **13.86 ns** | **0.2175 ns** | **0.2035 ns** | **13.87 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |   96 |  True | 15.93 ns | 0.2334 ns | 0.2069 ns | 15.99 ns |  1.15 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **128** | **False** | **14.61 ns** | **0.2589 ns** | **0.2295 ns** | **14.58 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  128 | False | 12.15 ns | 0.2984 ns | 0.3552 ns | 12.04 ns |  0.83 |    0.03 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **128** |  **True** | **14.25 ns** | **0.1148 ns** | **0.1073 ns** | **14.26 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  128 |  True | 15.23 ns | 0.1533 ns | 0.1280 ns | 15.21 ns |  1.07 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **160** | **False** | **15.70 ns** | **0.1604 ns** | **0.1422 ns** | **15.72 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  160 | False | 12.51 ns | 0.3498 ns | 0.4424 ns | 12.37 ns |  0.81 |    0.03 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **160** |  **True** | **15.82 ns** | **0.0845 ns** | **0.0749 ns** | **15.81 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  160 |  True | 15.70 ns | 0.3088 ns | 0.2888 ns | 15.70 ns |  0.99 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **192** | **False** | **16.30 ns** | **0.0710 ns** | **0.0554 ns** | **16.31 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  192 | False | 12.59 ns | 0.0563 ns | 0.0499 ns | 12.59 ns |  0.77 |    0.00 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **192** |  **True** | **16.10 ns** | **0.1212 ns** | **0.1133 ns** | **16.13 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  192 |  True | 16.06 ns | 0.2854 ns | 0.2530 ns | 16.01 ns |  1.00 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **224** | **False** | **17.85 ns** | **0.1759 ns** | **0.1559 ns** | **17.80 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  224 | False | 12.86 ns | 0.2408 ns | 0.2134 ns | 12.81 ns |  0.72 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **224** |  **True** | **17.83 ns** | **0.2890 ns** | **0.2256 ns** | **17.84 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  224 |  True | 17.93 ns | 0.5082 ns | 0.4753 ns | 17.78 ns |  1.01 |    0.03 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **256** | **False** | **19.24 ns** | **0.3690 ns** | **0.3452 ns** | **19.10 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  256 | False | 14.90 ns | 0.3580 ns | 0.3349 ns | 14.80 ns |  0.77 |    0.02 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **256** |  **True** | **19.49 ns** | **0.3181 ns** | **0.2976 ns** | **19.43 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  256 |  True | 18.02 ns | 0.5497 ns | 0.4873 ns | 17.91 ns |  0.93 |    0.03 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **999** | **False** | **36.22 ns** | **0.3804 ns** | **0.3372 ns** | **36.15 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  999 | False | 20.43 ns | 0.4568 ns | 0.4273 ns | 20.36 ns |  0.56 |    0.01 |
|               |      |       |          |           |           |          |       |         |
| **OldLengthCtor** |  **999** |  **True** | **37.19 ns** | **0.3663 ns** | **0.3248 ns** | **37.17 ns** |  **1.00** |    **0.00** |
| NewLengthCtor |  999 |  True | 29.86 ns | 0.5559 ns | 0.4642 ns | 29.77 ns |  0.80 |    0.02 |

Reply:
Going one step further to isolate the change and only updating the array if `defaultValue=true`, the difference in using Fill vs. looping for `defaultValue=false` case disappears (obviously, since neither code executes in that case).

> Should that be reverted, or are the cases different somehow?

We use fill in the ctor but not in SetAll because:
> 2. Given the ctor is doing more work (in all cases) rather than just calling Fill, the regression for small arrays isn't that extreme (20% instead of 3x).

Should we avoid it in the ctor too (due to the regression for small inputs)?

Reply:
> Should we avoid it in the ctor too

Do we have any data on common sizes used with BitArray?

Reply:
> Do we have any data on common sizes used with BitArray?

I browsed the dotnet and aspnet repos to see how BitArray is used.

It looks like 1-4 ints is common, i.e. 32-128 bits (for instance in EF).

- [Over 32](https://github.com/aspnet/Routing/blob/bd481034fea82bc48e1c1153e2e1a4f066012375/src/Microsoft.AspNetCore.Routing/Matching/CandidateSet.cs#L121)
- [64](https://github.com/aspnet/EntityFramework6/blob/22a11d132306c257435fa7d6d75a31359cbbd05a/src/EntityFramework/Core/Query/InternalTrees/VarVec.cs#L311)
- [384/767](https://github.com/dotnet/roslyn/blob/fab7134296816fc80019c60b0f5bef7400cf23ea/src/Workspaces/Core/Portable/Shared/Utilities/BloomFilter.cs#L45)
- [< 64](https://github.com/dotnet/machinelearning-archive/blob/fc4b6d727e912952df0a4fbaae58872deed08d61/src/Microsoft.ML.Transforms/NAReplaceTransform.cs#L218)
- [< 64](https://github.com/dotnet/machinelearning/blob/28fb3bbca4d5b8ad5b1f0bc28fef35db4d8232f3/src/Microsoft.ML.Transforms/MissingValueReplacing.cs#L260)

However, in ML, where they use it for feature.Count, it is generally large:
- [100 to 1000s](https://github.com/dotnet/machinelearning/blob/dbbc69edbc4ca05f7b67a30d82861da3ef2e2fc7/src/Microsoft.ML.Ensemble/Selector/FeatureSelector/RandomFeatureSelector.cs#L55)
- [100s](https://github.com/dotnet/machinelearning/blob/8a45f37cf87e380a93146d08acac19f215648f9a/src/Microsoft.ML.FastTree/FastTree.cs#L1728)


Also, seems like EF has their own copy of BitArray as well, for perf reasons. Maybe it is no longer necessary.
https://github.com/aspnet/EntityFramework6/blob/22a11d132306c257435fa7d6d75a31359cbbd05a/src/EntityFramework/Core/Query/PlanCompiler/Validator.cs#L316

However, I am not sure if we should worry too much about < 64. For anyone who cares about performance for < 64 and micro-optimizing it, I would assume they would use a ulong and only allocate for > 64 (like I do in the JsonReader). For that case, the overhead is minimal/disappears.

Thoughts?

Reply:
Ok. Thanks for looking.